### ISSUE TITLE
Opencv 3.2 failing to build with CUDA 8.0

### ISSUE THREAD

**Issue by nishantb21** [opencv/opencv#7992]
- OpenCV => 3.2.0 Release
- Operating System / Platform => Windows 10 64 bit
- Compiler => Visual Studio 2015 Community

##### Detailed description
I have searched on a lot of forums and no one seems to have a solution to this peculiar problem. The compilation of OpenCV 3.2 works fine unless WITH_CUDA is set to ON. I tried building OpenCV 3.2.0 with CUDA support, giving the appropriate architecture and giving the OPENCV_EXTRA_MODULES_PATH. While Cmake is able to both configure the project and generate it without any issues, the ALL_BUILD fails to build 4 files. Which inclusion of CUDA libraries works just fine, during runtime errors arise. The errors in particular are:
1.opencv\modules\core\include\opencv2/core/base.hpp(365): warning : function declared with "noreturn" does return
2.opencv\modules\core\include\opencv2/core/types.hpp(491): warning : field of class type without a DLL interface used in a class with a DLL interface

These are warnings, granted. However, these eventually lead to runtime errors while executing programs. One of these errors include the program unable to find entry points into cuda dlls.

**Comment by alalek** [comment]
1. False positive. error() throws an exception so it never returns.
2. Windows specific problem. Real problems are not observed yet.

>  these eventually lead to runtime errors while executing programs

What kind of errors?

**Comment by nishantb21** [comment]
I shall post the entire build log, the exact runtime errors and my source code I used for testing the library in a couple of hours. It's compiling right now and as usual its through a bunch of warnings which you, along with some other people, have attributed to the OS and not an inherent problem with the library.

**Comment by nishantb21** [comment]
Sorry for the delay in posting this. Just wanted to make sure I am not wasting anybody's time here.

So here are the build and install log files from the compilation of OpenCV using Visual Studio 2015. As you will see the build log for both Debug and Release configurations are pretty useless because like you pointed out, the problem is with the compiler on Windows platform. Interesting thing however comes to light when we inspect the install logs for Debug and Release configuration. As we see the error is as follows:

**1. error C2610: 'std::tuple<cv::Size,perf::`anonymous-namespace'::MatDepth,perf::`anonymous-namespace'::MatCn>::tuple(std::tuple<cv::Size,perf::`anonymous-namespace'::MatDepth,perf::`anonymous-namespace'::MatCn> &&)': is not a special member function which can be defaulted (compiling source file C:\Users\ANONYMOUS-PC\Desktop\CCBD Poject\cmake_opencv_cuda\opencv\modules\cudaarithm\perf\perf_reductions.cpp)**

**2. error C2535: 'std::tuple<cv::Size,perf::`anonymous-namespace'::MatDepth,perf::`anonymous-namespace'::MatCn> &std::tuple<cv::Size,perf::`anonymous-namespace'::MatDepth,perf::`anonymous-namespace'::MatCn>::operator =(const std::tuple<cv::Size,perf::`anonymous-namespace'::MatDepth,perf::`anonymous-namespace'::MatCn> &)': member function already defined or declared (compiling source file C:\Users\ANONYMOUS-PC\Desktop\CCBD Poject\cmake_opencv_cuda\opencv\modules\cudaarithm\perf\perf_reductions.cpp)**

Don't know if this is actually the issue or not. After all of this it says that opencv was compiled without CUDA support whenever I use the cv::cuda::setDevice function even though I clearly have enabled the WITH_CUDA flag. cv::cuda::getCudaEnabledDeviceCount returns 0 even though I have a GT 740m CUDA enabled GPU on my system. I have also tried this with a laptop containing a 960m and the build is failing there too.
[Install-log-debug.txt](https://github.com/opencv/opencv/files/704400/Install-log-debug.txt)
[Install-log-release.txt](https://github.com/opencv/opencv/files/704401/Install-log-release.txt)
[Build-log-debug.txt](https://github.com/opencv/opencv/files/704402/Build-log-debug.txt)
[Build-log-release.txt](https://github.com/opencv/opencv/files/704403/Build-log-release.txt)

**Comment by nishantb21** [comment]
This is the peculiar runtime error I was talking about! Hopefully you can shed some light on this!

![error](https://cloud.githubusercontent.com/assets/19548448/21954313/53df4b14-da73-11e6-99e3-5b36ce11cbcb.png)

**Comment by liukaimcu** [comment]
mute the thread
----- 原始邮件 -----
发件人：Nishant Bhattacharya <notifications@github.com>
收件人：opencv/opencv <opencv@noreply.github.com>
抄送人：Subscribed <subscribed@noreply.github.com>
主题：Re: [opencv/opencv] Opencv 3.2 failing to build with CUDA 8.0 (#7992)
日期：2017年01月14日 18点36分

This is the peculiar runtime error I was talking about! Hopefully you can shed some light on this!


&mdash;

**Comment by jasjuang** [comment]
I am also getting the same compilation error when building OpenCV from source on Windows 10 with Visual Studio 2015.

C2610 and C2535 error for file tuple in the opencv_perf_cudawarping project.
C2382 error for file cuda_perf.hpp in the opencv_perf_cudawarping project.

And lot more similar errors related to opencv_perf. Would be appreciated if this could be fixed0

A side question is is it possible to provide pre-built OpenCV on windows with CUDA? From my experience the compilation fail on windows happens way often then on Ubuntu, and I need the CUDA version of OpenCV for my code.

**Comment by nishantb21** [comment]
Okay so apparently if you switch off tests(both performance and normal) the error in the build log goes away. The .dll issue however persists. Still no solution for that.

**Comment by alalek** [comment]
@nishantb21 
Text error messages via screenshots must be avoided, because they are useless to search related information.

Please add these logs:
- CMake stdout (from clean build)
- files from build folder: CMakeCache.txt and CMakeVars.txt (as attachment or use "grep" with CUDA/CXX keywords)

Current MSVC error messages doesn't help a lot. There is related issue #6716

@jasjuang 
> provide pre-built OpenCV on windows with CUDA

There are several problems:
- CUDA library/runtime version dependency
- there is large set of possible GPU architectures, so resulting binary size is very large

**Comment by nishantb21** [comment]
Okay so a couple of updates on my part:
1. I tried compiling OpenCV on Ubuntu. Worked perfectly fine with CUDA 8.0, the same version which I was using on Windows. So as you pointed out there is definitely an issue with Windows.
2. As for the large set of possible GPU Architectures, I changed CUDA_ARCH to only 3.0 so it doesn't build for unnecessary architectures which my GPU doesn't support.
3. As for you suggested in that issue #6716, I will surely try disabling OPENCL and then trying. If it still exists I shall report back with the logs you requested.
Cheers!

**Comment by jasjuang** [comment]
@alalek I don't mind having to download a large binary at all. Downloading is way easier then figuring out various compilation error on windows. And yes the compilation error I am getting is exactly the same as what #6716 is showing.

**Comment by nishantb21** [comment]
Update 2:
1. I tried compiling with OpenCL switched off and it still didn't work. Same errors.
2. I am attaching the requested files so you maybe able to figure out what the error is!
[Cmake-stdout.txt](https://github.com/opencv/opencv/files/716793/Cmake-stdout.txt)
[CMakeVars.txt](https://github.com/opencv/opencv/files/716794/CMakeVars.txt)
[CMakeCache.txt](https://github.com/opencv/opencv/files/716795/CMakeCache.txt)

**Comment by alalek** [comment]
Thank you for update!

MSVS 2015 version is fresh (update 3), but there are [some reports](https://devtalk.nvidia.com/default/topic/883704/cuda-setup-and-installation/when-will-visual-studio-2015-be-supported-/post/4986548/#4986548) about unsupported compiler versions by CUDA. Please check you have the latest version of CUDA SDK.

BTW, please avoid using non alphanumeric symbols in paths (you have a space in "CCBD Poject").

**Comment by nishantb21** [comment]
Okay so I checked. I do happen to have the latest version CUDA SDK (v8.0.44, which is the same one I have on my Ubuntu machine). And as for the point about the inclusion of non alphanumeric characters, duly noted. So this error is because Visual Studio 2015 is misbehaving with the CUDA SDK? Well if that's the case then we cannot really do anything till the fix comes up I guess? One possible solution seems to be to VS 2017. But don't know if I will be able to try that out until later because I am quite busy. 
Thanks for the help anyways! Cheers!

**Comment by arpit1997** [comment]
> Windows specific problem. Real problems are not observed yet.
these eventually lead to runtime errors while executing programs

CUDA 8.0 does not have a cuda toolkit so it might be a problem however i got it working on my ubuntu machine with opencv 3.1 and cuda 8.0. Can't say for sure about windows.
:small_blue_diamond: :small_orange_diamond: :smile_cat:

**Comment by ThorbjornSomod** [comment]
@nishantb21: 

Have you gotten any further on this issue? I am getting the samme errors with the same setup (except im using a GTX960,  so 5.2 CC and Maxwell architecture in the CMake). I am trying to build OpenCV 3.2 with opencv_contrib  and CUDA support. The Cmake seems to configure and generate the solution just fine, but when I try to build the OpenCV.sln project I get the same "noreturn" and DLL warnings as you do. The solution completes it's build both in Debug and Release, but when I try to use the cuda modules in my example program I get an error saying that OpenCV was not built with CUDA support.

**Comment by nishantb21** [comment]
@ThorbjornSomod 
I am sorry to inform you that I have been really busy with work and have not gotten around to try and fix the issue. I haven't used OpenCV from the time of my last comment on this thread so I really don't know if the latest commits to this repo have fixed it or not. Could you tell me if you tried compiling on VS 2017? It seems to be a possible solution. The "noreturn" as @alalek mentioned is simply a warning which is not to be worried about. I am inclined to think that the latest NVidia drivers may be a possible issue but I am not sure. On Ubuntu 16.04 on which I installed OpenCV with CUDA support successfully without any errors, I used Nvidia Driver version 367.57 which is not the latest one. In fact some threads out there also tell you to NOT install the latest NVidia drivers as they have not been tested very well on Ubuntu. With Windows you could try installing an older driver and trying it but that would be a pain. I suggest you just natively install Ubuntu on your machine if you really want it that much and try compilation there. Cheers!

**Comment by blackTay** [comment]
I have the same problem (building OpenCV 3.2 with CUDA v8 for Win10). I asked on stackoverflow ([link](http://stackoverflow.com/questions/42122634/building-opencv-3-cuda-errors)).
Could it work with Cuda 7.5 instead of 8??

**Comment by ThorbjornSomod** [comment]
@nishantb21 & @blackTay : I actually got it to work both on my laptop and my desktop (GTX960M and GTX970 respectively) running with OpenCV 3.2 and the latest version of CUDA 8.0 for Win10 in Visual Studio 15 Community! What I did was to enable WITH_CUBLAS aswell as WITH_CUDA. I also turned off BUILD_PERF_TESTS and BUILD_TESTS. The configuration was built using the Visual Studio 14 2015 C++ compiler.

**Comment by blackTay** [comment]
@ThorbjornSomod Thank you! This is a good work-around until it's fixed. Do you know the reason for this problem?

**Comment by ThorbjornSomod** [comment]
I think that some version of BLAS (found [here](http://www.openblas.net/) is needed to correctly build the CUDA interface. If you enable WITH_BLAS you are basicly saying that you dont have BLAS installed natively and would like OpenCV to be build with their version of it. Or something like this?

**Comment by nishantb21** [comment]
@ThorbjornSomod So are you saying that to build OpenCV 3.2 with CUDA you either need BLAS installed on your system or check the WITH_BLAS flag so that OpenCV can build BLAS on your system? Well according to this reasoning shouldn't it only affect WITH_CUBLAS only and not WITH_CUDA? Also could you share your Cmake log files as I have shared on this forum before? I would like to see if any other settings need to be met by my build. I will try to replicate your environment and build OpenCV 3.2 with CUDA support on my system. Thank you!

**Comment by nishantb21** [comment]
Okay so I know its has been a couple of months since my last update but turns out this issue has been fixed. I just recently cloned the repo and tried building it on my new laptop. I got the same warnings as before. But somehow it seems to work just fine on my laptop now. The cudaEnabledDeviceCount returns 1. Finally. I am going to attach my new build logs for people who would like to investigate what was different from the point of view of my hardware and software. 
[CMakeCache.txt](https://github.com/opencv/opencv/files/952699/CMakeCache.txt)
[CMakeVars.txt](https://github.com/opencv/opencv/files/952700/CMakeVars.txt)
